content/en/blog/_posts/2021-09-20-admission-controllers-and-container-drift/index.md

```diff
@@ -0,0 +1,80 @@
+
+Here comes a sequence diagram of the entire workflow mentioned above: 
+{{< figure src="workflow-diagram.png" alt="Workflow Diagram" >}}
+
+## New kubectl plugin for a better user experience
+Our admission controller service works great for solving the container drift issue we had on the platform. It also submits related K8s events to the target pod that has been affected. However, as K8s events get away quickly (one hour of persistency by default configuration), we need to provide other mechanisms for developers to get their pod interaction activity. A [kubectl plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) is a perfect choice for us to expose this information. We named our plugin “kubectl-pi” (short for pod-interaction) and provided two subcommands: “get” and “extend”. When the “get” command is called, the plugin simply checks the label attached by our admission controller service and transfers it to human-readable info. Here is an example output of the `kubectl pi get` command:
+
+```
+$ kubectl pi get test-pod
+POD_NAME    INTERACTOR    POD_TTL    EXTENSION    EXTENSION_REQUESTER    EVICTION_TIME
+test-pod    username-1    1h0m0s     /            /                      2021-09-20 11:00:00 -0800 PST
+```
+
+The plugin can also be used to extend the current eviction time of a pod, in case developers need more buffer to debug ongoing issues after exec’ing into their containers. This is achieved by the `kubectl pi extend` command, where the plugin patches relevant K8s annotation to the given pod. This annotation includes the duration and username who made the extension request for transparency (displayed in the table returned from “get” command). There is another webhook defined in our admission controller service that admits this annotation request and resets the eviction timer of the target pod accordingly. An example of requesting the extension from the developer side would be:
```

Comment:
```suggestion
The plugin can also be used to extend the current eviction time of a pod, in case developers need more buffer to debug ongoing issues after executing on their containers. This is achieved by the `kubectl pi extend` command, where the plugin patches relevant K8s annotation to the given pod. This annotation includes the duration and username who made the extension request for transparency (displayed in the table returned from “get” command). There is another webhook defined in our admission controller service that admits this annotation request and resets the eviction timer of the target pod accordingly. An example of requesting the extension from the developer side would be:
```